File: platform/backend/app/api/events.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2026 CommonHuman-Lab
from datetime import datetime
from typing import Any, Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, field_validator
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db, require_admin
from app.models.ctf_event import EventScoringMode, EventStatus, EventVisibility
from app.models.user import User
from app.services import audit_service
from app.services.event_service import (
    add_challenge, get_event_by_slug_or_404, get_event_challenges,
    get_event_or_404, is_scoreboard_frozen, list_events,
    register_team, remove_challenge, transition_status, unregister_team,
)
from app.services.scoring_service import get_event_scoreboard

router = APIRouter(prefix="/events", tags=["events"])


# ── Schemas ───────────────────────────────────────────────────────────────────

class EventSummary(BaseModel):
    id: int
    slug: str
    title: str
    description: Optional[str]
    status: EventStatus
    visibility: EventVisibility
    scoring_mode: EventScoringMode
    start_at: Optional[datetime]
    end_at: Optional[datetime]
    max_team_size: Optional[int]
    freeze_scoreboard_at: Optional[datetime]
    created_at: datetime
    scoreboard_frozen: bool = False

    model_config = {"from_attributes": True}


class EventCreateRequest(BaseModel):
    slug: str
    title: str
    description: Optional[str] = None
    start_at: Optional[datetime] = None
    end_at: Optional[datetime] = None
    visibility: EventVisibility = EventVisibility.PRIVATE
    scoring_mode: EventScoringMode = EventScoringMode.STATIC
    max_team_size: Optional[int] = None
    freeze_scoreboard_at: Optional[datetime] = None


class EventUpdateRequest(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    start_at: Optional[datetime] = None
    end_at: Optional[datetime] = None
    visibility: Optional[EventVisibility] = None
    max_team_size: Optional[int] = None
    freeze_scoreboard_at: Optional[datetime] = None


class AddChallengeRequest(BaseModel):
    challenge_id: int
    points_override: Optional[int] = None
    released_at: Optional[datetime] = None


class RegisterTeamRequest(BaseModel):
    team_id: int


class TransitionRequest(BaseModel):
    status: EventStatus


class ScoreboardEntry(BaseModel):
    rank: int
    user_id: Optional[int] = None
    team_id: Optional[int] = None
    total: int
    last_tx: Optional[str] = None


def _serialize(ev) -> EventSummary:
    return EventSummary(
        id=ev.id,
        slug=ev.slug,
        title=ev.title,
        description=ev.description,
        status=ev.status,
        visibility=ev.visibility,
        scoring_mode=ev.scoring_mode,
        start_at=ev.start_at,
        end_at=ev.end_at,
        max_team_size=ev.max_team_size,
        freeze_scoreboard_at=ev.freeze_scoreboard_at,
        created_at=ev.created_at,
        scoreboard_frozen=is_scoreboard_frozen(ev),
    )


# ── Public / authenticated endpoints ─────────────────────────────────────────

@router.get("/", response_model=list[EventSummary])
def list_events_endpoint(
    status: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[EventSummary]:
    include_private = current_user.is_admin or current_user.is_superuser
    events = list_events(db, status=status, include_private=include_private)
    return [_serialize(ev) for ev in events]


@router.get("/{slug}", response_model=EventSummary)
def get_event_endpoint(
    slug: str,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
) -> EventSummary:
    ev = get_event_by_slug_or_404(db, slug)
    return _serialize(ev)


@router.get("/{slug}/challenges", response_model=list[dict[str, Any]])
def get_event_challenges_endpoint(
    slug: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[dict[str, Any]]:
    ev = get_event_by_slug_or_404(db, slug)
    return get_event_challenges(db, ev, user_id=current_user.id)


@router.get("/{slug}/scoreboard", response_model=list[ScoreboardEntry])
def event_scoreboard_endpoint(
    slug: str,
    limit: int = Query(100, ge=1, le=500),
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
) -> list[ScoreboardEntry]:
    ev = get_event_by_slug_or_404(db, slug)
    rows = get_event_scoreboard(db, event_id=ev.id, limit=limit)
    return [ScoreboardEntry(**r) for r in rows]


@router.post("/{slug}/register", response_model=dict[str, Any])
def register_endpoint(
    slug: str,
    body: RegisterTeamRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict[str, Any]:
    ev = get_event_by_slug_or_404(db, slug)
    reg = register_team(db, ev, body.team_id)
    return {"event_id": ev.id, "team_id": reg.team_id, "registered_at": reg.registered_at.isoformat()}


@router.delete("/{slug}/register/{team_id}", status_code=204)
def unregister_endpoint(
    slug: str,
    team_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
) -> None:
    ev = get_event_by_slug_or_404(db, slug)
    unregister_team(db, ev, team_id)


# ── Admin-only endpoints ──────────────────────────────────────────────────────

@router.post("/", response_model=EventSummary, dependencies=[Depends(require_admin)])
def create_event_endpoint(
    body: EventCreateRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> EventSummary:
    from app.models.ctf_event import CtfEvent
    ev = CtfEvent(
        slug=body.slug,
        title=body.title,
        description=body.description,
        start_at=body.start_at,
        end_at=body.end_at,
        visibility=body.visibility,
        scoring_mode=body.scoring_mode,
        max_team_size=body.max_team_size,
        freeze_scoreboard_at=body.freeze_scoreboard_at,
        created_by_id=current_user.id,
    )
    db.add(ev)
    db.commit()
    db.refresh(ev)
    audit_service.write_audit(db, "event.created", user_id=current_user.id, detail={"slug": ev.slug})
    return _serialize(ev)


@router.patch("/{slug}", response_model=EventSummary, dependencies=[Depends(require_admin)])
def update_event_endpoint(
    slug: str,
    body: EventUpdateRequest,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
) -> EventSummary:
    ev = get_event_by_slug_or_404(db, slug)
    if body.title is not None:
        ev.title = body.title
    if body.description is not None:
        ev.description = body.description
    if body.start_at is not None:
        ev.start_at = body.start_at
    if body.end_at is not None:
        ev.end_at = body.end_at
    if body.visibility is not None:
        ev.visibility = body.visibility
    if body.max_team_size is not None:
        ev.max_team_size = body.max_team_size
    if body.freeze_scoreboard_at is not None:
        ev.freeze_scoreboard_at = body.freeze_scoreboard_at
    db.commit()
    db.refresh(ev)
    return _serialize(ev)


@router.post("/{slug}/status", response_model=EventSummary, dependencies=[Depends(require_admin)])
def transition_status_endpoint(
    slug: str,
    body: TransitionRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> EventSummary:
    ev = get_event_by_slug_or_404(db, slug)
    ev = transition_status(db, ev, body.status)
    audit_service.write_audit(
        db, "event.status_changed",
        user_id=current_user.id,
        detail={"slug": ev.slug, "new_status": ev.status.value},
    )
    return _serialize(ev)


@router.post("/{slug}/challenges", response_model=dict[str, Any], dependencies=[Depends(require_admin)])
def add_challenge_endpoint(
    slug: str,
    body: AddChallengeRequest,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
) -> dict[str, Any]:
    ev = get_event_by_slug_or_404(db, slug)
    ecm = add_challenge(db, ev, body.challenge_id, body.points_override, body.released_at)
    return {
        "event_id": ev.id,
        "challenge_id": ecm.challenge_id,
        "points_override": ecm.points_override,
        "released_at": ecm.released_at.isoformat() if ecm.released_at else None,
    }


@router.delete("/{slug}/challenges/{challenge_id}", status_code=204, dependencies=[Depends(require_admin)])
def remove_challenge_endpoint(
    slug: str,
    challenge_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
) -> None:
    ev = get_event_by_slug_or_404(db, slug)
    remove_challenge(db, ev, challenge_id)

File: platform/frontend/lib/api/events.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 CommonHuman-Lab
import { apiClient } from "./client";
import type { ScoreboardEntry, ChallengeDifficulty } from "./challenges";

export type EventStatus = "draft" | "published" | "running" | "ended" | "archived";
export type EventVisibility = "public" | "private" | "unlisted";
export type EventScoringMode = "static" | "dynamic";

export interface CtfEvent {
  id: number;
  slug: string;
  title: string;
  description: string | null;
  status: EventStatus;
  visibility: EventVisibility;
  scoring_mode: EventScoringMode;
  start_at: string | null;
  end_at: string | null;
  max_team_size: number | null;
  freeze_scoreboard_at: string | null;
  created_at: string;
  scoreboard_frozen: boolean;
}

export interface EventChallenge {
  id: number;
  slug: string;
  title: string;
  category: string;
  difficulty: ChallengeDifficulty;
  points: number;
  tags: string[];
  solve_count: number;
  solved_by_me: boolean;
  released_at: string | null;
}

export async function getEvents(status?: string): Promise<CtfEvent[]> {
  const { data } = await apiClient.get<CtfEvent[]>("/events/", {
    params: status ? { status } : {},
  });
  return data;
}

export async function getEvent(slug: string): Promise<CtfEvent> {
  const { data } = await apiClient.get<CtfEvent>(`/events/${slug}`);
  return data;
}

export async function getEventChallenges(slug: string): Promise<EventChallenge[]> {
  const { data } = await apiClient.get<EventChallenge[]>(`/events/${slug}/challenges`);
  return data;
}

export async function getEventScoreboard(
  slug: string,
  limit = 100
): Promise<ScoreboardEntry[]> {
  const { data } = await apiClient.get<ScoreboardEntry[]>(
    `/events/${slug}/scoreboard`,
    { params: { limit } }
  );
  return data;
}

export async function registerTeam(slug: string, teamId: number): Promise<void> {
  await apiClient.post(`/events/${slug}/register`, { team_id: teamId });
}

export interface CreateEventPayload {
  slug: string;
  title: string;
  description?: string;
  start_at?: string;
  end_at?: string;
  visibility: EventVisibility;
  scoring_mode: EventScoringMode;
  max_team_size?: number;
}

export async function createEvent(payload: CreateEventPayload): Promise<CtfEvent> {
  const { data } = await apiClient.post<CtfEvent>("/events/", payload);
  return data;
}

export async function transitionEvent(slug: string, status: EventStatus): Promise<CtfEvent> {
  const { data } = await apiClient.post<CtfEvent>(`/events/${slug}/status`, { status });
  return data;
}

export interface UpdateEventPayload {
  title?: string;
  description?: string | null;
  start_at?: string | null;
  end_at?: string | null;
  visibility?: EventVisibility;
  scoring_mode?: EventScoringMode;
  max_team_size?: number | null;
  freeze_scoreboard_at?: string | null;
}

export async function updateEvent(slug: string, payload: UpdateEventPayload): Promise<CtfEvent> {
  const { data } = await apiClient.patch<CtfEvent>(`/events/${slug}`, payload);
  return data;
}

export async function addEventChallenge(
  slug: string,
  challengeId: number,
  pointsOverride?: number | null
): Promise<void> {
  await apiClient.post(`/events/${slug}/challenges`, {
    challenge_id: challengeId,
    points_override: pointsOverride ?? null,
  });
}

export async function removeEventChallenge(slug: string, challengeId: number): Promise<void> {
  await apiClient.delete(`/events/${slug}/challenges/${challengeId}`);
}
